### ISSUE TITLE
EmailAddress.objects.add_email is still not atomic [emails are case insensitive]

### ISSUE THREAD

**Issue by jleclanche** [pennersr/django-allauth#1873]
I thought I had fixed the race condition in #1711 (36f4862066223a3ddd3c1bbfaa199cc13d0dbc21). Unfortunately, the following code still has a race condition:

```py
email_address, created = self.get_or_create(
    user=user, email__iexact=email, defaults={"email": email}
)
```

The problem I missed is that there is no unique-together on user+email **when the email is specified as a different case**. 
Furthermore, while debugging this, I found that the `app_settings.UNIQUE_EMAIL` account setting removes the unique_together if set to False. This is incorrect -- UNIQUE_EMAIL enforces uniqueness of the `email` field, but has nothing to do with the uniqueness *per-user*.

My suggestion:

1. Lowercase the `email` field whenever it's saved to the database.
2. Add a migration and/or a management command to lowercase and cleanup all duplicate emails.
3. Remove the `if not app_settings.UNIQUE_EMAIL:` condition from the unique_together.

There are a lot of iexact throughout the allauth code which would benefit from having database-enforced lowercase emails as well.

**Comment by lukeburden** [comment]
Separate to the uniqueness enforcement/race-condition above, there's another reason to consider moving towards storing all email values in lower case.

The `iexact` lookups on both the `EmailAddress` and `User` models scale linearly as the table size increases, as the database must compute `UPPER` for each row in the table. Once you have a decent number of users, this becomes a DOS vulnerability in `allauth`. I've seen different projects using allauth end up having their database CPU pegged at 100% when scripters inevitably try password brute force attacks.

**Comment by lukeburden** [comment]
But it does feel challenging to move away from the current `iexact`-heavy approach. I think the steps @jleclanche suggest make sense, but we'd also need to consider the `email` field on the `User` model and how we can ensure it is also lower cased.

**Comment by jplehmann** [comment]
We also have noted that the emails are case-sensitive which makes the lookups we are wanting to do difficult (there is no `iin__=`).

**Comment by michelts** [comment]
Maybe we can take the same approach @pennersr [took with the username](https://github.com/pennersr/django-allauth/commit/7538d907aad3eabb749ed3cd74dd8fea6e46a473), I mean, we could add a settings entry ACCOUNT_PRESERVE_EMAIL_CASING that defaults to true, or we can simply rely on the ACCOUNT_PRESERVE_USERNAME_CASING, because as I could understand, this problem happens for email-based authentication (e.g. ACCOUNT_USER_MODEL_USERNAME_FIELD is None).

**Comment by derek-adair** [comment]
emails should 100% be lowercase.  No more settings frivolous settings... right....................... right? :dancers:

**Comment by lukeburden** [comment]
Several years on, just looping back to retract the additional issue I highlighted above - the database inefficiency of doing iexact on large user and email tables. This was resolved with the recent addition of [an appropriate index](https://github.com/pennersr/django-allauth/commit/b072cca54695237e5174dffa52db6d1b5eade23c). That just leaves the possible race condition due to the get_or_create not being keyed on an index.

**Comment by pennersr** [comment]
> emails should 100% be lowercase

Not according to the standards. The part before the `@` is not guaranteed to be case insensitive. So formally, the right thing to do is to store the email as typed, yet, for duplicate checks use an insensitive comparison -- which is what allauth is doing. 

For reference: https://stackoverflow.com/questions/9807909/are-email-addresses-case-sensitive

**Comment by pennersr** [comment]
Having said that, I am leaning towards deviating from the standard and always storing and looking up email addresses in a lower case fashion. Rationale:

- A cursory analysis of others shows they are storing emails as lower case as well (Keycloak, Ory, Salesforce, ...).
- Performance issues cannot be solved on some databases (#3453)

**Comment by pennersr** [comment]
See: https://github.com/pennersr/django-allauth/pull/3460 -- would appreciate some eyes on that PR (@lukeburden?)

**Comment by derek-adair** [comment]
> > emails should 100% be lowercase
> 
> Not according to the standards. 

WOW.  did NOT know this.  Looks like django usernames are case insensitive by default as well!  Seems severely problematic for any public facing application.  I'm in a bit of disbelief this is a thing.

Also FWIW [django supports functional unique constraints as of 4.0](https://docs.djangoproject.com/en/4.1/releases/4.0/#functional-unique-constraints).  Something I will be doing on my user models going forward! Yeeeesh.

**Comment by pennersr** [comment]
Closed via #3460

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/1873
